opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -12,7 +12,214 @@
+MAX_NUM_ANNOTATIONS = 32
+MAX_NUM_EVENTS = 128
+MAX_NUM_LINKS = 32
+
+AttributeValue = typing.Union[str, bool, float]
+
+
+class BoundedList(Sequence):
+    """An append only list with a fixed max size."""
+    def __init__(self, maxlen):
+        self.dropped = 0
+        self._dq = deque(maxlen=maxlen)
+        self._lock = threading.Lock()
+
+    def __repr__(self):
+        return ("{}({}, maxlen={})"
+                .format(
+                    type(self).__name__,
+                    list(self._dq),
+                    self._dq.maxlen
+                ))
+
+    def __getitem__(self, index):
+        return self._dq[index]
+
+    def __len__(self):
+        return len(self._dq)
+
+    def __iter__(self):
+        return iter(self._dq)
+
+    def append(self, item):
+        with self._lock:
+            if len(self._dq) == self._dq.maxlen:
+                self.dropped += 1
+            self._dq.append(item)
+
+    def extend(self, seq):
+        with self._lock:
+            to_drop = len(seq) + len(self._dq) - self._dq.maxlen
+            if to_drop > 0:
+                self.dropped += to_drop
+            self._dq.extend(seq)
+
+    @classmethod
+    def from_seq(cls, maxlen, seq):
+        seq = tuple(seq)
+        if len(seq) > maxlen:
+            raise ValueError
+        bounded_list = cls(maxlen)
+        bounded_list._dq = deque(seq, maxlen=maxlen)
+        return bounded_list
+
+
+class BoundedDict(MutableMapping):
+    """A dict with a fixed max capacity."""
+    def __init__(self, maxlen):
+        self.maxlen = maxlen
+        self.dropped = 0
+        self._dict = OrderedDict()
+        self._lock = threading.Lock()
+
+    def __repr__(self):
+        return ("{}({}, maxlen={})"
+                .format(
+                    type(self).__name__,
+                    dict(self._dict),
+                    self.maxlen
+                ))
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __setitem__(self, key, value):
+        with self._lock:
+            if key in self._dict:
+                del self._dict[key]
+            elif len(self._dict) == self.maxlen:
+                del self._dict[next(iter(self._dict.keys()))]
+                self.dropped += 1
+            self._dict[key] = value
+
+    def __delitem__(self, key):
+        del self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    @classmethod
+    def from_map(cls, maxlen, mapping):
+        mapping = OrderedDict(mapping)
+        if len(mapping) > maxlen:
+            raise ValueError
+        bounded_dict = cls(maxlen)
+        bounded_dict._dict = mapping
+        return bounded_dict
+
+
+class SpanContext(trace_api.SpanContext):
+    """See `opentelemetry.trace.SpanContext`."""
+
+    def is_valid(self) -> bool:
+        return (self.span_id == trace_api.INVALID_SPAN_ID or
+                self.trace_id == trace_api.INVALID_SPAN_ID)
```

Comment:
`self.trace_id == trace_api.INVALID_TRACE_ID`

Reply:
I don't have strong opinion here, perhaps in the code logic we want to always put `trace_id` before `span_id` for consistency?

Reply:
Good call.